homeassistant/components/geocaching/oauth.py

```diff
@@ -0,0 +1,82 @@
+"""oAuth2 functions and classes for Geocaching API integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any, cast
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import DOMAIN, OAUTH2_AUTHORIZE_URL, OAUTH2_TOKEN_URL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class GeocachingOAuth2Implementation(
+    config_entry_oauth2_flow.LocalOAuth2Implementation
+):
+    """Local OAuth2 implementation for Geocaching."""
+
+    def __init__(
+        self, hass: HomeAssistant, client_id: str, client_secret: str, name: str
+    ):
+        """Local Geocaching Oauth Implementation."""
+        self._name = name
+        super().__init__(
+            hass=hass,
+            client_id=client_id,
+            client_secret=client_secret,
+            domain=DOMAIN,
+            authorize_url=OAUTH2_AUTHORIZE_URL,
+            token_url=OAUTH2_TOKEN_URL,
+        )
+
+    @property
+    def name(self) -> str:
+        """Name of the implementation."""
+        return f"{self._name}"
+
+    @property
+    def extra_authorize_data(self) -> dict:
+        """Extra data that needs to be appended to the authorize url."""
+        data = {"scope": "*", "response_type": "code"}
+        return data
+
+    async def async_resolve_external_data(self, external_data: Any) -> dict:
+        """Initialize local Geocaching API auth implementation."""
+        redirect_uri = external_data["state"]["redirect_uri"]
+        data = {
+            "grant_type": "authorization_code",
+            "code": external_data["code"],
+            "redirect_uri": redirect_uri,
+        }
+        token = await self._token_request(data)
+        # Store the redirect_uri (Needed for refreshing token, but not according to oAuth2 spec!)
+        token["redirect_uri"] = redirect_uri
+        return token
+
+    async def _async_refresh_token(self, token: dict) -> dict:
+        """Refresh tokens."""
+        data = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "grant_type": "refresh_token",
+            "refresh_token": token["refresh_token"],
+            # Add previously stored redirect_uri (Mandatory, but not according to oAuth2 spec!)
```

Comment:
Can you report this as a bug?

Reply:
Not sure if passing extra data to the refresh token request is currently supported. @bramkragten ? If it is, it should whitelist keys on a per domain basis.

Reply:
> Can you report this as a bug?

I have sent a mail to my contact person at Geocaching. They haven't respond to it, although my mail is from a few months ago. I guess they won't change this because they've documented it here so it looks to me it was on purpose.

https://api.groundspeak.com/documentation#authorization

Reply:
I've had the pleasure of more OAuth services not according to spec (e.g., Toon). Maybe we need OAuth quirks :S

Reply:
I'll get to this next week

Reply:
@bramkragten, we have official confirmation from Geocaching that our integration can use the production API.  Paulus has the production keys for this API  in his mailbox. If I can be of any help, please let me know!

Reply:
@bramkragten @balloob Any news on this? would be nice if we could still get this in the beta :)

Reply:
It is in the pipeline, but there is another change blocking this at this point.
We know it won't make it for 2021.7

Reply:
Ah ok, thanks for the heads up, then we have to be patient 😄

Reply:
Any update on this?

Reply:
@balloob ?

Reply:
hi @bramkragten, welcome back from holiday! Why is this PR still open? Is there something I can do to speed up the process?

Reply:
@bramkragten @balloob Can somebody tell me why this PR still isn't merged? I really want to move forward to prevent Geocaching to revoke our API access.

Reply:
I don't think our cloud has been updated yet to support the extra keys.

Reply:
Hi @balloob @bramkragten, any update on this one?

Reply:
And we are on our way! 🎉